web3-functions/pyth-oracle-w3f-priceIds/index.ts

```diff
@@ -0,0 +1,235 @@
+    } catch (err) {
+      const error = err as Error;
+      return {
+        canExec: false,
+        message: `Error fetching gist: ${error.message}`,
+      };
+    }
+  } else {
+    console.log("using pythConfig from storage");
+    pythConfig = pythConfigStorage.pythConfig;
+  }
+
+  console.log(`pythConfig: ${JSON.stringify(pythConfig, null, 2)}`);
+
+  const {
+    pythNetworkAddress,
+    priceServiceEndpoint,
+    validTimePeriodSeconds,
+    deviationThreshold,
+    priceIds,
+  } = pythConfig;
+
+  const pythnetwork = new Contract(
+    pythNetworkAddress,
+    pythAbi,
+    provider
+  ) as IPyth;
+
+  // Get Pyth price data
+  const connection = new EvmPriceServiceConnection(priceServiceEndpoint); // See Price Service endpoints section below for other endpoints
+
+  console.log(`checking priceIds: ${priceIds}`);
+  const check = (await connection.getLatestPriceFeeds(priceIds)) as any[];
+  console.log(`check: ${JSON.stringify(check, null, 2)}`);
+
+  const currentPrices: Map<string, IPRICE> = new Map();
+  check
+    .filter((c) => {
+      return c.price && c.price.price;
+    })
+    .forEach((c) => {
+      console.log(`c: ${JSON.stringify(c, null, 2)}`);
+      const price = {
+        price: +c.price.price,
+        timestamp: +c.price.publishTime,
+      };
+      currentPrices.set(addLeading0x(c.id), price);
+    });
+
+  console.log(
+    `currentPrices: ${JSON.stringify([...currentPrices.entries()], null, 2)}`
+  );
+
+  if (currentPrices.size != priceIds.length) {
+    console.log(
+      `currentPrices.priceIds: ${JSON.stringify(currentPrices.keys())}`
+    );
+    return { canExec: false, message: "Not all prices available" };
+  }
+
+  const lastPrices = (
+    await Promise.all(
+      [...currentPrices.keys()].map(async (priceId) => {
+        const price = JSON.parse(
+          (await storage.get(priceId)) ?? "{}"
+        ) as IPRICE;
+        return { priceId, price };
+      })
+    )
+  )
+    .filter((p) => p.price.price != undefined)
+    .reduce((acc, priceInfo) => {
+      acc.set(priceInfo.priceId, priceInfo.price);
+      return acc;
+    }, new Map<string, IPRICE>());
+
+  console.log(
+    `lastPrices: ${JSON.stringify([...lastPrices.entries()], null, 2)}`
+  );
+
+  const priceIdsToInitialize = [...currentPrices.keys()].filter((priceId) => {
+    return lastPrices.get(priceId) == undefined;
+  });
+
+  const priceIdsToUpdate = [...currentPrices.keys()]
+    .filter((priceId) => lastPrices.get(priceId) != undefined)
+    .filter((priceId) => {
+      const lastPrice = lastPrices.get(priceId)!;
+      const currentPrice = currentPrices.get(priceId)!;
+      const priceDiff =
+        Math.abs(lastPrice.price - currentPrice.price) / lastPrice.price;
+      const priceExceedsDiff = priceDiff >= deviationThreshold;
+      const priceIsStale =
+        currentPrice.timestamp - lastPrice.timestamp > validTimePeriodSeconds;
+      return priceExceedsDiff || priceIsStale;
+    });
```

Comment:
seems like you could simplify the logic here by making priceIdsToUpdate include the ids where `lastPrices.get(priceId) === undefined`. Then you wouldn't need `priceIdsToInitialize`.